### ISSUE TITLE
hostnamed: Assertion '(_error) != 0' failed at src/hostname/hostnamed.c:207, function fallback_chassis(). Aborting.

### ISSUE THREAD

**Issue by gyakovlev** [systemd/systemd#20160]
**systemd version the issue has been seen with**
 >  systemd 249 running in system mode (+PAM -AUDIT -SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN -IPTC +KMOD -LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY -P11KIT +QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified)

<!-- **NOTE:** Do not submit bug reports about anything but the two most recently released *major* systemd versions upstream! -->
<!--           If there have been multiple stable releases for that major version, please consider updating to a recent one before reporting an issue. -->
<!--           When using a distro package, please make sure that the version reported is meaningful for upstream. -->
<!-- See https://github.com/systemd/systemd-stable/releases for the list of most recent releases. -->
<!-- For older version please use distribution trackers (see https://systemd.io/CONTRIBUTING#filing-issues). -->

**Used distribution**
 > gentoo

**Linux kernel version used** (`uname -a`)
<!-- Make sure to enclose the pasted kernel version in `backticks`, so that
     GitHub doesn't convert the `#` character typically included in it into a
     reference to old GitHub issues. -->
 > `Linux cerberus 5.10.47 #1 SMP Fri Jul 2 12:55:24 PDT 2021 ppc64le POWER9, altivec supported PowerNV C1P9S01 REV 1.01 GNU/Linux`

**CPU architecture issue was seen on**
 > ppc64le

**Expected behaviour you didn't see**
 > no crash. returns data.

**Unexpected behaviour you saw**
 > hostnamectl 
 > Failed to query system properties: Remote peer disconnected

**Steps to reproduce the problem**
 > start hostnamed, run hostnamectl. hostnamed crashes.

**Additional program output to the terminal or log subsystem illustrating the issue**
```
× systemd-hostnamed.service - Hostname Service                                                                                                                                                                                                         [0/267]
     Loaded: loaded (/usr/lib/systemd/system/systemd-hostnamed.service; static)
     Active: failed (Result: core-dump) since Wed 2021-07-07 15:30:45 PDT; 2min 24s ago
       Docs: man:systemd-hostnamed.service(8)
             man:hostname(5)
             man:machine-info(5)
             man:org.freedesktop.resolve1(5)
    Process: 66233 ExecStart=/usr/lib/systemd/systemd-hostnamed (code=dumped, signal=ABRT)
   Main PID: 66233 (code=dumped, signal=ABRT)
        CPU: 211ms

Jul 07 15:30:45 cerberus systemd[1]: Starting Hostname Service...
Jul 07 15:30:45 cerberus systemd[1]: Started Hostname Service.
Jul 07 15:30:45 cerberus systemd-hostnamed[66233]: Assertion '(_error) != 0' failed at src/hostname/hostnamed.c:207, function fallback_chassis(). Aborting.
Jul 07 15:30:45 cerberus systemd[1]: systemd-hostnamed.service: Main process exited, code=dumped, status=6/ABRT
Jul 07 15:30:45 cerberus systemd[1]: systemd-hostnamed.service: Failed with result 'core-dump'.
```
<!-- Please paste relevant program terminal or journal output here, ideally
     when generated in debug mode (try setting the SYSTEMD_LOG_LEVEL=debug
     environment variable). For very long copy/pasted data consider using a
     service like https://gist.github.com/. Where copy/paste is not possible
     (for example early boot or late shutdown), a photo of the screen might do
     too, but text is always much preferred. -->
```gdb
           PID: 66233 (systemd-hostnam)
           UID: 0 (root)
           GID: 0 (root)
        Signal: 6 (ABRT)
     Timestamp: Wed 2021-07-07 15:30:45 PDT (7min ago)
  Command Line: /usr/lib/systemd/systemd-hostnamed
    Executable: /usr/lib/systemd/systemd-hostnamed
 Control Group: /system.slice/systemd-hostnamed.service
          Unit: systemd-hostnamed.service
         Slice: system.slice
       Boot ID: 383c476ee92d4cd79300c7c43bae5402
    Machine ID: 061a262264704116bbbcdbe578d9cadb
      Hostname: cerberus
       Storage: /var/lib/systemd/coredump/core.systemd-hostnam.0.383c476ee92d4cd79300c7c43bae5402.66233.1625697045000000.zst (present)
     Disk Size: 143.3K
       Message: Process 66233 (systemd-hostnam) of user 0 dumped core.
                
                Found module libz.so.1 without build-id.
                Found module libgpg-error.so.0 without build-id.
                Found module ld64.so.2 without build-id.
                Found module libpthread.so.0 without build-id.
                Found module libm.so.6 without build-id.
                Found module liblzma.so.5 without build-id.
                Found module libzstd.so.1 without build-id.
                Found module libseccomp.so.2 without build-id.
                Found module librt.so.1 without build-id.
                Found module libpam.so.0 without build-id.
                Found module libcrypto.so.1.1 without build-id.
                Found module libmount.so.1 without build-id.
                Found module liblz4.so.1 without build-id.
                Found module libkmod.so.2 without build-id.
                Found module libgcrypt.so.20 without build-id.
                Found module libdl.so.2 without build-id.
                Found module libcrypt.so.2 without build-id.
                Found module libcap.so.2 without build-id.
                Found module libblkid.so.1 without build-id.
                Found module libacl.so.1 without build-id.
                Found module libc.so.6 without build-id.
                Found module libsystemd-shared-249.so without build-id.
                Found module systemd-hostnamed without build-id.
                Stack trace of thread 66233:
                #0  0x00003fffbbe756dc __GI_raise (libc.so.6 + 0x456dc)
                #1  0x00003fffbbe561fc __GI_abort (libc.so.6 + 0x261fc)
                #2  0x00003fffbc20ba78 log_assert_failed (libsystemd-shared-249.so + 0x1aba78)
                #3  0x0000000108f13738 fallback_chassis (systemd-hostnamed + 0x3738)
                #4  0x0000000108f15564 context_fallback_icon_name (systemd-hostnamed + 0x5564)
                #5  0x0000000108f15610 property_get_icon_name (systemd-hostnamed + 0x5610)
                #6  0x00003fffbc27e6c8 invoke_property_get (libsystemd-shared-249.so + 0x21e6c8)
                #7  0x00003fffbc27f120 vtable_append_one_property (libsystemd-shared-249.so + 0x21f120)
                #8  0x00003fffbc280488 vtable_append_all_properties (libsystemd-shared-249.so + 0x220488)
                #9  0x00003fffbc281d5c property_get_all_callbacks_run (libsystemd-shared-249.so + 0x221d5c)
                #10 0x00003fffbc282c9c bus_process_object (libsystemd-shared-249.so + 0x222c9c)
                #11 0x00003fffbc29c3d4 process_message (libsystemd-shared-249.so + 0x23c3d4)
                #12 0x00003fffbc29d200 io_callback (libsystemd-shared-249.so + 0x23d200)
                #13 0x00003fffbc30e010 source_dispatch (libsystemd-shared-249.so + 0x2ae010)
                #14 0x00003fffbc30e3b4 sd_event_dispatch (libsystemd-shared-249.so + 0x2ae3b4)
                #15 0x00003fffbc30ecc8 sd_event_run (libsystemd-shared-249.so + 0x2aecc8)
                #16 0x00003fffbc0fd338 bus_event_loop_with_idle (libsystemd-shared-249.so + 0x9d338)
                #17 0x0000000108f15124 run (systemd-hostnamed + 0x5124)
                #18 0x0000000108f13258 main (systemd-hostnamed + 0x3258)
                #19 0x00003fffbbe56634 generic_start_main (libc.so.6 + 0x26634)
                #20 0x00003fffbbe56820 __libc_start_main (libc.so.6 + 0x26820)
```

**Comment by gyakovlev** [comment]
I can do a debug build if needed, but here's bt full as is

```gdb
(gdb) bt full           
#0  0x00003fffbbe756dc in raise () from /usr/lib64/libc.so.6
No symbol table info available.
#1  0x00003fffbbe561fc in abort () from /usr/lib64/libc.so.6
No symbol table info available.
#2  0x00003fffbc20ba78 in log_assert_failed (text=<optimized out>, file=<optimized out>, line=<optimized out>, func=<optimized out>) at ../systemd-249/src/basic/log.c:862
No locals.
#3  0x0000000108f13738 in fallback_chassis () at ../systemd-249/src/hostname/hostnamed.c:207
        _error = 0
        type = 0x0
        t = 0
        v = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = "fallback_chassis"
        __func__ = "fallback_chassis"
#4  0x0000000108f15564 in context_fallback_icon_name (c=<optimized out>) at ../systemd-249/src/hostname/hostnamed.c:314
        chassis = <optimized out>
        __PRETTY_FUNCTION__ = "context_fallback_icon_name"
#5  context_fallback_icon_name (c=<optimized out>) at ../systemd-249/src/hostname/hostnamed.c:306
        chassis = <optimized out>
        __PRETTY_FUNCTION__ = "context_fallback_icon_name"
#6  0x0000000108f15610 in property_get_icon_name (bus=<optimized out>, path=<optimized out>, interface=<optimized out>, property=<optimized out>, reply=0x13de990d0, userdata=0x3fffca8dee50, error=<optimized out>) at ../systemd-249/src/hostname/hostnamed.c:692
        n = 0x0
        c = 0x3fffca8dee50
        name = <optimized out>
#7  0x00003fffbc27e6c8 in invoke_property_get (bus=bus@entry=0x13de94260, slot=0x13de96340, v=v@entry=0x108f2f538 <hostname_vtable+336>, path=path@entry=0x13de98cd8 "/org/freedesktop/hostname1", interface=0x13de963c0 "org.freedesktop.hostname1", property=0x108f180d0 "IconName", reply=reply@entry=0x13de990d0, userdata=0x3fffca8dee50, error=error@entry=0x3fffca8de870) at ../systemd-249/src/libsystemd/sd-bus/bus-objects.c:456
        p = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = "invoke_property_get"
#8  0x00003fffbc27f120 in vtable_append_one_property (bus=bus@entry=0x13de94260, reply=reply@entry=0x13de990d0, path=path@entry=0x13de98cd8 "/org/freedesktop/hostname1", c=c@entry=0x13de96378, v=v@entry=0x108f2f538 <hostname_vtable+336>, userdata=userdata@entry=0x3fffca8dee50, error=error@entry=0x3fffca8de870) at ../systemd-249/src/libsystemd/sd-bus/bus-objects.c:738
        slot = 0x13de96340
        r = <optimized out>
        __PRETTY_FUNCTION__ = "vtable_append_one_property"
#9  0x00003fffbc280488 in vtable_append_all_properties (bus=bus@entry=0x13de94260, reply=0x13de990d0, path=0x13de98cd8 "/org/freedesktop/hostname1", c=c@entry=0x13de96378, userdata=0x3fffca8dee50, error=error@entry=0x3fffca8de870) at ../systemd-249/src/libsystemd/sd-bus/bus-objects.c:794
        v = 0x108f2f538 <hostname_vtable+336>
        r = <optimized out>
        __PRETTY_FUNCTION__ = "vtable_append_all_properties"
#10 0x00003fffbc281d5c in property_get_all_callbacks_run (found_object=0x3fffca8de9c7, iface=<optimized out>, require_fallback=false, first=<optimized out>, m=0x13de98d80, bus=0x13de94260) at ../systemd-249/src/libsystemd/sd-bus/bus-objects.c:855
        error = {name = 0x0, message = 0x0, _need_free = 0}
        u = 0x3fffca8dee50
        reply = 0x13de990d0
        c = 0x13de96378
        found_interface = true
        r = <optimized out>
        reply = <optimized out>
        c = <optimized out>
        found_interface = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = {<optimized out> <repeats 31 times>}
        error = {name = <optimized out>, message = <optimized out>, _need_free = <optimized out>}
        u = <optimized out>
#11 object_find_and_run (bus=bus@entry=0x13de94260, m=m@entry=0x13de98d80, p=<optimized out>, require_fallback=require_fallback@entry=false, found_object=found_object@entry=0x3fffca8de9c7) at ../systemd-249/src/libsystemd/sd-bus/bus-objects.c:1369
        iface = 0x0
        get = <optimized out>
        n = 0x13de962f0
        vtable_key = {path = 0x13de98cd8 "/org/freedesktop/hostname1", interface = 0x13de98d10 "org.freedesktop.DBus.Properties", member = 0x13de98d00 "GetAll", parent = 0x3fffca8de990, last_iteration = 30000000, vtable = 0x3fffca8de940}
        v = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = "object_find_and_run"
#12 0x00003fffbc282c9c in bus_process_object (bus=0x13de94260, m=0x13de98d80) at ../systemd-249/src/libsystemd/sd-bus/bus-objects.c:1443
        prefix = 0x13de980a0 "\230\336\023"
        r = <optimized out>
        pl = <optimized out>
        found_object = true
        __PRETTY_FUNCTION__ = "bus_process_object"
#13 0x00003fffbc29c3d4 in process_message (m=0x13de98d80, bus=0x13de94260) at ../systemd-249/src/libsystemd/sd-bus/sd-bus.c:2955
        r = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = {<optimized out> <repeats 16 times>}
        __func__ = "process_message"
        finish = <optimized out>
        _mm = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
#14 process_running (ret=0x0, bus=0x13de94260) at ../systemd-249/src/libsystemd/sd-bus/sd-bus.c:2997
        m = 0x13de98d80
        r = 1
        m = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = {<optimized out> <repeats 16 times>}
        null_message = <optimized out>
        __func__ = "process_running"
        _found = <optimized out>
        __assert_in_set = {<optimized out>, <optimized out>}
        _ptr_ = <optimized out>
        _level = <optimized out>
        _e = <optimized out>
#15 bus_process_internal (bus=0x13de94260, ret=0x0) at ../systemd-249/src/libsystemd/sd-bus/sd-bus.c:3217
        r = <optimized out>
        __PRETTY_FUNCTION__ = "bus_process_internal"
        _dont_destroy_bus = 0x13de94260
#16 0x00003fffbc29d200 in io_callback (s=<optimized out>, fd=<optimized out>, revents=<optimized out>, userdata=<optimized out>) at ../systemd-249/src/libsystemd/sd-bus/sd-bus.c:3595
        bus = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = "io_callback"
        __func__ = "io_callback"
#17 0x00003fffbc30e010 in source_dispatch (s=0x13de97b50) at ../systemd-249/src/libsystemd/sd-event/sd-event.c:3504
        saved_event = 0x13de92ae0
        saved_type = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = "source_dispatch"
        __func__ = "source_dispatch"
#18 0x00003fffbc30e3b4 in sd_event_dispatch (e=0x13de92ae0) at ../systemd-249/src/libsystemd/sd-event/sd-event.c:4078
        ref = 0x13de92ae0
        p = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = "sd_event_dispatch"
#19 0x00003fffbc30ecc8 in sd_event_run (e=0x13de92ae0, timeout=30000000) at ../systemd-249/src/libsystemd/sd-event/sd-event.c:4139
        r = <optimized out>
        __PRETTY_FUNCTION__ = "sd_event_run"
        ref = 0x13de92ae0
#20 0x00003fffbc0fd338 in bus_event_loop_with_idle (e=0x13de92ae0, bus=0x13de94260, name=0x108f17c78 "org.freedesktop.hostname1", timeout=30000000, check_idle=0x0, userdata=0x0) at ../systemd-249/src/shared/bus-util.c:104
        idle = <optimized out>
        exiting = false
        r = <optimized out>
        code = 0
        __PRETTY_FUNCTION__ = "bus_event_loop_with_idle"
#21 0x0000000108f15124 in run (argc=<optimized out>, argv=<optimized out>) at ../systemd-249/src/hostname/hostnamed.c:1275
        context = {data = {0x13de99770 "cerberus", 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, hostname_source = HOSTNAME_STATIC, etc_hostname_stat = {st_dev = 29, st_ino = 327766, st_nlink = 1, st_mode = 33188, st_uid = 0, st_gid = 0, __pad2 = 0, st_rdev = 0, st_size = 9, st_blksize = 512, st_blocks = 2, st_atim = {tv_sec = 1625642816, tv_nsec = 26784643}, st_mtim = {tv_sec = 1622754198, tv_nsec = 111578670}, st_ctim = {tv_sec = 1622754198, tv_nsec = 112578535}, __glibc_reserved4 = 0, __glibc_reserved5 = 0, __glibc_reserved6 = 0}, etc_os_release_stat = {st_dev = 0, st_ino = 0, st_nlink = 0, st_mode = 0, st_uid = 0, st_gid = 0, __pad2 = 0, st_rdev = 0, st_size = 0, st_blksize = 0, st_blocks = 0, st_atim = {tv_sec = 0, tv_nsec = 0}, st_mtim = {tv_sec = 0, tv_nsec = 0}, st_ctim = {tv_sec = 0, tv_nsec = 0}, __glibc_reserved4 = 0, __glibc_reserved5 = 0, __glibc_reserved6 = 0}, etc_machine_info_stat = {st_dev = 0, st_ino = 0, st_nlink = 0, st_mode = 0, st_uid = 0, st_gid = 0, __pad2 = 0, st_rdev = 0, st_size = 0, st_blksize = 0, st_blocks = 0, st_atim = {tv_sec = 0, tv_nsec = 0}, st_mtim = {tv_sec = 0, tv_nsec = 0}, st_ctim = {tv_sec = 0, tv_nsec = 0}, __glibc_reserved4 = 0, __glibc_reserved5 = 0, __glibc_reserved6 = 0}, polkit_registry = 0x0}
        event = 0x13de92ae0
        bus = 0x13de94260
        r = <optimized out>
        __PRETTY_FUNCTION__ = "run"
        __func__ = "run"
#22 0x0000000108f13258 in main (argc=<optimized out>, argv=<optimized out>) at ../systemd-249/src/hostname/hostnamed.c:1282
        r = <optimized out>
```

**Comment by gyakovlev** [comment]
adding `CHASSIS=desktop` to `/etc/machine-id` fixes the crash, as it no longer attempts to guess the type and takes different codepath.

**Comment by gyakovlev** [comment]
realized I forgot to include host info.
It's bare-metal PowerNV machine, built on Raptor Blackbird motherboard.

**Comment by floppym** [comment]
Should be fixed by #20159.
